.github/workflows/main.yml

```diff
@@ -33,8 +33,30 @@ jobs:
       if: matrix.compiler == 'clang' || matrix.compiler == 'clang-cl'
       shell: bash
       run: |
+        set -euo pipefail
         if [ "${OSTYPE}" = "msys" ]; then
-          choco install --no-progress llvm
+          export MSYS2_ARG_CONV_EXCL="*"  # Don't let MSYS2 attempt to auto-translate arguments that look like paths
+          # Ideally we should be able to use the Chocolatey package manager:
+          #   choco install --no-progress llvm
+          # However, it frequently gives HTTP 503 errors, so we just download and install manually.
+          urldir="https://releases.llvm.org"
+          bits=64
+          if [ "${HOSTTYPE}" = "${HOSTTYPE%64}" ]; then bits=32; fi
```

Comment:
Let's just download `http://releases.llvm.org/9.0.0/LLVM-9.0.0-win64.exe` here? There is no need to support 32 bit architectures and also it is better to be explicit wrt. the version, because if it gets upgraded, we want to upgrade it manually to make sure the build doesn't break.

Reply:
@pcmoritz Regarding 32-bit, I don't know how widely it's used; I just figured adding an extra line for it wasn't a big deal.
Regarding downloading a specific version—I did think about that, but that would make the code inconsistent across platforms. Right now we don't have any control over what version is used on Linux or Mac (and I'm not sure we ever will—and neither would we if we did `choco install`), so if we pin this to a specific version, then Windows behavior would get out-of-sync. I'm not sure what the value gain from pinning a version would be if we did that; if anything I'd expect it to be more of a headache later when something e.g. fails to compile on Windows but compiles fine on other platforms.

Reply:
The scenario that we should try to avoid is that something breaks without us changing the Ray sourcecode, i.e. if LLVM makes a new release and then suddenly the windows CI breaks. That's why all libraries are pinned to a specific version/commit. It is fine if windows uses a different version of LLVM than linux I think, as long as they are both work.

What do you think about that?

Reply:
Oh I don't generally disagree with it being good to pin to a specific version. I'm just concerned if it's done inconsistently, it'll be more of a headache to figure out later, and I'm worried if the error isn't obvious, it's likely to result in builds getting merged that then actually *do* break on Windows. So on my end I'd prefer to do this for all platforms (preferably in another PR). But if you still really want me to special-case Windows behavior, I'll do it... though in that case, then let me know where/how to specify the version? (Unless you're fine with it being buried in the code here?)

Reply:
Ah ok, I'm not sure if I follow! So the scenario I'm concerned is the following:

We use the latest LLVM version, there is an update of LLVM that has a backward incompatible change that for some reason we depended on. It automatically gets used in the windows CI, which as a result fails.

As opposed to:

There is a backward incompatible change and we want to upgrade. Somebody has to create a PR that bumps the version and the tests in the PR fail. It can be fixed in the PR and master is not affected.

Imagine Ray would break every time on of the libraries we depend on made a backward incompatible change. That would lead to a lot of sadness.

I'm not sure if I understand your concern, if somebody makes a change that breaks windows, the CI will detect it and the PR shouldn't get merged, right?

The version of LLVM we use could be specified in an environment variable in this file, that seems like a good way to do it.

Reply:
Thanks for explaining that :) I think you're right—thinking about it again, this is unlikely to occur due to _compilability_ issues specifically.

Here's another kind of scenario I can imagine though, where CI won't detect it necessarily:

C++ sometimes adds rules that affect run-time behaviors, e.g. by defining previously-undefined behavior. (This has happened a few times, such as when they introduced a new threading model, and when they constrained the order of sub-expression evaluation.)  
What happens when a new version of some library starts to depend on such a constraint for correctness? It's very likely to get merged into master, because the specific edge cases that exhibit it can easily never occur in CI (either due to lack of coverage or due to nondeterminism). So it'll silently break master on Windows (causing silent data corruption, etc.) for a long time, until/unless somebody updates the Windows compiler, intentionally or accidentally. It'll be much, _much_ harder for us to track down, since it'll look like a Windows-specific issue. Whereas if the compilers on all three platforms get updated simultaneously, it'd either never occur in the first place, or if so, it'd get noticed and tracked down much sooner (both because non-Windows users will see it too, and because we wouldn't start from the false assumption that it's Windows-related).

So that's one reason why I'd prefer to keep the compilers in lockstep with each other—broadly, to avoid non-platform-specific issues from misleading and confusing us into thinking they're platform-specific.

Another concern I have is simpler: that even if an issue does show up in CI, I can imagine the PR getting merged in despite breaking Windows builds—perhaps because the cause isn't obvious (possibly because it, again, looks like a Windows-specific issue). So this will again break master (and risk more code divergence and incompatibilities during that entire period) until someone upgrades the compiler.

So it seems like a trade-off to me: Do we prefer the potential of having Windows broken on master (and potentially across releases) for a longer period of time, or do we prefer the potential of having _all_ platforms broken on master for a much shorter period of time?  
Which one is a higher risk to take? Honestly I don't know for sure. Personally, I also prefer to avoid introducing new degrees of freedom (variability across platforms, which can cause confusion and complexity), and I really prefer to get issues fixed as quickly as possible for users, even if it means blocking the entire team—at least (or especially) if the issues might potentially include something like data corruption for some subset of users.  
But it depends a lot on how likely you think each scenario is, and how easy do you think it will be to resolve things in each case. So if you feel the trade-off isn't worth it, that's fine, I can make the change anyway. :) And honestly, I can see this is more a longer-term risk to worry about, so if we pin to a specific compiler version on Linux and Mac in the foreseeable future, problems likely won't come up at all.

Reply:
I pinned the LLVM version on Windows, should be hopefully ready to merge.